src/types/quota.rs

```diff
@@ -0,0 +1,231 @@
+use crate::error::{Error, ParseError};
+use crate::parse::{parse_many_into, parse_many_into2, MapOrNot, MapOrNot2};
+use crate::types::UnsolicitedResponse;
+use imap_proto::Response;
+use ouroboros::self_referencing;
+use std::borrow::Cow;
+use std::fmt::{Debug, Display, Formatter};
+use std::sync::mpsc;
+
+/// From [SETQUOTA Resource limit](https://datatracker.ietf.org/doc/html/rfc2087#section-4.1)
+///
+/// Used by [`Session::set_quota`]
+pub struct QuotaResourceLimit<'a> {
+    /// The resource type
+    pub name: QuotaResourceName<'a>,
+    /// The amount for that resource
+    pub amount: u64,
+}
+
+impl Display for QuotaResourceLimit<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.name, self.amount)
+    }
+}
+
+/// From [Resources](https://datatracker.ietf.org/doc/html/rfc2087#section-3)
+///
+/// Used by [`QuotaLimit`], and [`QuotaResource`]
+#[derive(Debug, Eq, PartialEq)]
+pub enum QuotaResourceName<'a> {
+    /// Sum of messages' RFC822.SIZE, in units of 1024 octets
+    Storage,
+    /// Number of messages
+    Message,
+    /// Any other string (for future RFCs)
+    Atom(Cow<'a, str>),
+}
+
+impl Display for QuotaResourceName<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Storage => write!(f, "STORAGE"),
+            Self::Message => write!(f, "MESSAGE"),
+            Self::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Due to an implementation detail this is a wrapper around [`Quota`] with the same contract.
+///
+/// Used by [`Session::get_quota`] and [`Session::set_quota`]
+#[self_referencing]
+pub struct QuotaResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: Quota<'this>,
+}
+
+impl QuotaResponse {
+    /// Parse the [`Quota`] response from a response buffer.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaResponseTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut quotas = Vec::new();
+
+                // There should only be one single QUOTA response
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
+                    Response::Quota(q) => Ok(MapOrNot::Map(Quota::from_imap_proto(q))),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                match quotas.len() {
+                    1 => Ok(quotas.remove(0)),
+                    _ => Err(Error::Parse(ParseError::Invalid(input.to_vec()))),
+                }
```

Comment:
I'd still prefer to see this pattern encapsulated as a `parse_until_done` helper that calls `parse_many_into` but expects to only find a single elements. It could also probably use `Option` instead of `Vec` as the `IT` type.

Reply:
Once we get everything resolved over in the ACL PR, I'll update this one to reflect the final decisions there.